### ISSUE TITLE
Enable IRQ sharing to remove the current limit of 11 devices per microVM

### ISSUE THREAD

**Issue by dhrgit** [firecracker-microvm/firecracker#1268]
Currently, Firecracker supports emulating only 11 virtio devices on x86_64, since each device gets assigned its own IRQ. This issue will track adding support for sharing a single IRQ across multiple devices, thus removing the limit imposed on the number of emulated devices.

**Comment by iabestever** [comment]
My partner and I will be working on this issue

**Comment by iabestever** [comment]
Hi @dhrgit,

Hope all is well. Just checking, did you get our email asking for more direction on this issue?

**Comment by dhrgit** [comment]
@iabestever The next step here was to test the performance impact of IRQ sharing, in order to determine if a round-robin assignment was good enough.

IRQ sharing may cause extra VM exists upon interrupt delivery, since multiple drivers may end up querying their respective devices for their interrupt status. These queries are done via MMIO reads, and every read incurs a VM exit.

For instance, if multiple high-throughput virtio-net devices would share an interrupt, every IRQ delivery would incur some extra VM exists, which might have some impact on performance (i.e. throughput, frames per second). No idea on the magnitude of that impact, though, so this scenario should be tested.

So, as a first step, a straight forward approach would be to iperf a few virtio-net devices, under two different scenarios: 1) each device gets its own IRQ, and 2) all devices share an IRQ, and see when / if throughput decays.

**Comment by iabestever** [comment]
Thank you @dhrgit, we'll get right on it

**Comment by iabestever** [comment]
![deviceOnIrq](https://user-images.githubusercontent.com/5136247/69923928-d671e280-146d-11ea-8f57-da53db7d5e3a.png)
@dhrgit so we ran iperf3 tests to see the throughput of virtio net devices. As we increased the number of virtio net devices, we ensured they were all assigned to the same IRQ. As you can see, as the numbers of devices on an IRQ increases the throughput gradually decreases. We did this be making x TAP interfaces and then attaching them to a microVM as described in #711.  

To combat the decrease in throughput, should we devise an irq affinity scheme that ensures network interfaces actively avoid sharing IRQs with other network interfaces? Or alternatively, is there a way to see which interfaces are high-load and which interfaces are low-load over the lifecycle of a microVM and then reassign IRQs based on that?

**Comment by dhrgit** [comment]
If I'm reading this correctly, your tests show the throughput maxing out at about 6Gbps. This sounds a bit low. Can you please post the full test setup (i.e. vm configs, hardware used, kernel versions, etc)?

Also, it looks like there's a higher throughput for two devices vs one. That sounds a bit counter-intuitive. Can you please post your data samples? Thanks!

**Comment by iabestever** [comment]
[hardware.txt](https://github.com/firecracker-microvm/firecracker/files/3915541/hardware.txt)
^This file is my hardware information.
[runscript.txt](https://github.com/firecracker-microvm/firecracker/files/3915543/runscript.txt)
^This file has the API commands we used to attach the virtio network devices to the microVM.

Here is the kernel I am using: Linux iabestever-Inspiron-N5110 5.0.0-36-generic #39~18.04.1-Ubuntu SMP Tue Nov 12 11:09:50 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux 

Here is the qemu version I have: QEMU emulator version 2.11.1(Debian 1:2.11+dfsg-1ubuntu7.20)

I think the higher throughput for two devices vs one may have been an outlier/an expression of the limitation of my hardware. I also only ran once for one device, so maybe more samples for one and two devices may yield better, more intuitive results. Here is the data sample list for your reference:
[data.xlsx](https://github.com/firecracker-microvm/firecracker/files/3915644/data.xlsx)

**Comment by dhrgit** [comment]
Thanks for the detailed information! Quick question, thought: for what are you using qemu in these tests?

Also, with Firecracker, please keep in mind that all emulation runs on a single thread. So, even without IRQ sharing, throughput is likely to drop with the number of net devices being emulated. What we're looking for is the difference in throughput drop for the two difference scenarios: sharing vs dedicated IRQs.

**Comment by iabestever** [comment]
In the experiments I performed, only one device was being emulated at any given time. So I would attach x amount of network interfaces to a microVM and then inside the microVM I would set each of them to be listening on a unique IP address. However, from outside the microVM, when doing the tests, at any given time, only network packets were being sent to only one IP address. I believe that should only be testing having network devices on a dedicated IRQ and the throughout decrease due to the increased VM exits as devices increase. Please correct me if I am wrong. 

So the first IRQ is taken up by the root block device - IRQ 5 in Intel. Then I devised a test such that there was one network device to one IRQ. Since there are 10 IRQs available, I ran this experiment 10 times. In essence, I did throughput testing with one network device to one IRQ. Then I did another test where I had two network devices, each with their own dedicated IRQ. I tested the throughput of both devices individually - i.e. the throughput of network device one indpendently of the throughput of network device two.  Then I did repeated this test with three network devices, each with their own dedicated IRQs. I repeated this for 10 network devices as we only have 10 IRQs. At each point, I found the throughput to remain practically constant (only minor variations). And this number is ~5.6Gbps. So, with 10 network interfaces with dedicated IRQs the throughput for each interface is ~5.6Gbps. However, if 10 network interfaces all shared one IRQ, then the throughput for each interface is ~3.519Gbps.     

I hope this helps. Please let me know if my understanding is unclear.

**Comment by dhrgit** [comment]
It makes sense for the throughput to drop even if only one device is under load, since every interrupt will incur a VM exit for each attached device, regardless of the device being idle or not. Hence the drop from 5.6Gbps to 3.6Gbps you observed.

While the above is a very interesting result, it would also be helpful to have the devices loaded concurrently and see the difference in the compound throughput, though. I.e. compare these two scenarios: 1) N devices sharing an IRQ are loaded with traffic at the same time, and 2) N devices, each with their own dedicated IRQ, are loaded with traffic at the same time.

Also, since there appears to be a considerable throughput drop, it would be interesting to know if the overall performance is more impacted by pairing similar or different devices on a single IRQ. I.e. if multiple net devices share an IRQ, there's a throughput drop; but, if a net and a block device share an IRQ, how would that impact their compound performance (throughput for net, IOPS + throughput for block).

To sum it up: we have three types of virtio deivces: block, net, and vsock. If two devices need to share an IRQ, we are interested in seeing which pairing options (e.g. net + block on one IRQ, or block + vsock on one IRQ, etc) have the least impact on the overall performance, when the devices that are sharing an IRQ are driven to high load.

**Comment by iabestever** [comment]
Sounds good. I will have performance results for the different device type matchups to you by Friday. 

https://www.thegeekdiary.com/how-to-create-virtual-block-device-loop-device-filesystem-in-linux/ Does this website accurately define how to create a virtual block device? I was following these steps and just want to confirm if this is indeed the correct way. 

Also, I realise Firecracker only has the option to attach devices before boot. May I ask why is that? Additionally will it be possible to reassign IRQs based on measuring load? Say further down the line we have arrived at the conclusion that pairing a network device and a vsock on one IRQ yields the least amount of of overall impact on performance. However, during runtime we see that we have paired a high load network device and a high load vsock. Say another IRQ has a low load vsock and a low load network device. In that scenario, can we, during runtime, reassign IRQs such that a high load network device is paired with a low load vsock?

**Comment by dhrgit** [comment]
Firecracker doesn't currently support any hotplug mechanism (e.g. PCI). All device configuration, including the assigned IRQ, is passed to the kernel at boot time only, via the kernel command line. This is why it isn't possible to reassign IRQs, nor attach new devices at runtime.

**Comment by luminitavoicu** [comment]
Closing this, as Firecracker now supports 19 devices per microVM, which is enough for the customer use cases. Please reopen if you have any additional comments.
